File: src/main/java/p00/Problem_0014.java
package p00;

public class Problem_0014 {

	public static long solve() {

		int longest_chain = 0;
		long best_number = 0;

		int[] chain_lengths = new int[1000000];

		for (int i = 1; i < 1000000; i++) {
			long number = i;
			int chain = 0;

			while (number != 1) {

				if (number < 1000000 && chain_lengths[(int) number] != 0) {
					chain += chain_lengths[(int) number];
					break;
				}

				chain++;

				if (number % 2 == 0)
					number /= 2;
				else {
					number = (number * 3 + 1) / 2;

					chain++;
				}

			}

			chain_lengths[i] = chain;

			if (chain > longest_chain) {
				best_number = i;
				longest_chain = chain;
			}

		}

		return best_number;
	}

}

File: src/main/kotlin/p00/Problem0014.kt
package p00

class Problem0014 {

    fun solve(): Long {
        var longestChain = 0
        var bestNumber: Long = 0
        val chainLengths = IntArray(1000000)
        for (i in 1..999999) {
            var number = i.toLong()
            var chain = 0
            while (number != 1L) {
                if (number < 1000000 && chainLengths[number.toInt()] != 0) {
                    chain += chainLengths[number.toInt()]
                    break
                }
                chain++
                if (number % 2 == 0L) number /= 2 else {
                    number = (number * 3 + 1) / 2
                    chain++
                }
            }
            chainLengths[i] = chain
            if (chain > longestChain) {
                bestNumber = i.toLong()
                longestChain = chain
            }
        }

        return bestNumber
    }

}